File: test/org/stringtemplate/test/TestGroupSyntax.java
package org.stringtemplate.test;

import org.junit.Test;
import static org.junit.Assert.assertEquals;
import org.stringtemplate.*;

public class TestGroupSyntax extends BaseTest {
    @Test public void testSimpleGroup() throws Exception {
        String templates =
            "t() ::= <<foo>>" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = STGroup.loadGroup(tmpdir+"/"+"t.stg");
        String expected =
            "t() ::= <<" + Misc.newline+
            "foo" + Misc.newline+
            ">>"+ Misc.newline;
        String result = group.show();
        assertEquals(expected, result);
    }

    @Test public void testMultiTemplates() throws Exception {
        String templates =
            "ta() ::= \"[<it>]\"" + Misc.newline +
            "duh() ::= <<hi there>>" + Misc.newline +
            "wow() ::= <<last>>" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = STGroup.loadGroup(tmpdir+"/"+"t.stg");
        String expected =
            "ta() ::= <<" +Misc.newline+
            "[<it>]" +Misc.newline+
            ">>" +Misc.newline+
            "duh() ::= <<" +Misc.newline+
            "hi there" +Misc.newline+
            ">>" +Misc.newline+
            "wow() ::= <<" +Misc.newline+
            "last" +Misc.newline+
            ">>"+ Misc.newline;
        String result = group.show();
        assertEquals(expected, result);
    }

    @Test public void testSingleTemplateWithArgs() throws Exception {
        String templates =
            "t(a,b) ::= \"[<a>]\"" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = STGroup.loadGroup(tmpdir+"/"+"t.stg");
        String expected =
            "t(a,b) ::= <<" + Misc.newline+
            "[<a>]" + Misc.newline+
            ">>"+ Misc.newline;
        String result = group.show();
        assertEquals(expected, result);
    }

    @Test public void testDefaultValues() throws Exception {
        String templates =
            "t(a={def1},b=\"def2\") ::= \"[<a>]\"" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = STGroup.loadGroup(tmpdir+"/"+"t.stg");
        String expected =
            "t(a={def1},b=\"def2\") ::= <<" + Misc.newline+
            "[<a>]" + Misc.newline+
            ">>"+ Misc.newline;
        String result = group.show();
        assertEquals(expected, result);
    }

    @Test public void testDefaultValueTemplateWithArg() throws Exception {
        String templates =
            "t(a={x | 2*<x>}) ::= \"[<a>]\"" + Misc.newline;

        writeFile(tmpdir, "t.stg", templates);
        STGroup group = STGroup.loadGroup(tmpdir+"/"+"t.stg");
        String expected =
            "t(a={x | 2*<x>}) ::= <<" + Misc.newline+
            "[<a>]" + Misc.newline+
            ">>"+ Misc.newline;
        String result = group.show();
        assertEquals(expected, result);
    }

	@Test public void testNestedTemplateInGroupFile() throws Exception {
		String templates =
			"t(a) ::= \"<a:{x | <x:{<it>}>}>\"" + Misc.newline;

		writeFile(tmpdir, "t.stg", templates);
		STGroup group = STGroup.loadGroup(tmpdir+"/"+"t.stg");
		String expected =
			"t(a) ::= <<\n" +
			"<a:{x | <x:{<it>}>}>\n" +
			">>"+ Misc.newline;
		String result = group.show();
		assertEquals(expected, result);
	}

	@Test public void testNestedDefaultValueTemplate() throws Exception {
		String templates =
			"t(a={x | <x:{<it>}>}) ::= \"ick\"" + Misc.newline;

		writeFile(tmpdir, "t.stg", templates);
		STGroup group = STGroup.loadGroup(tmpdir+"/"+"t.stg");
		String expected =
			"t(a={x | <x:{<it>}>}) ::= <<\n" +
			"ick\n" +
			">>"+ Misc.newline;
		String result = group.show();
		assertEquals(expected, result);
	}

	@Test public void testNestedDefaultValueTemplateWithEscapes() throws Exception {
		String templates =
			"t(a={x | \\< <x:{<it>\\}}>}) ::= \"[<a>]\"" + Misc.newline;

		writeFile(tmpdir, "t.stg", templates);
		STGroup group = STGroup.loadGroup(tmpdir+"/"+"t.stg");
		String expected =
			"t(a={x | \\< <x:{<it>\\}}>}) ::= <<" + Misc.newline+
			"[<a>]" + Misc.newline+
			">>"+ Misc.newline;
		String result = group.show();
		assertEquals(expected, result);
	}
}

File: objc/main/tests/objc/TestGroupSyntax.m
#import "TestGroupSyntax.h"

@implementation TestGroupSyntax

- (void) testSimpleGroup
{
    NSString *templates = @"t() ::= <<foo>>\n";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    NSString *expected = @"t() ::= <<\nfoo\n>>\n";
    NSString *result = [group show];
    [self assertEquals:expected result:result];
}

- (void) testMultiTemplates
{
    NSString *templates = @"ta(x) ::= \"[<x>]\"\nduh() ::= <<hi there>>\nwow() ::= <<last>>\\n";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    NSString *expected = @"ta(x) ::= <<\n[<x>]\n>>\nduh() ::= <<\nhi there\n>>\nwow() ::= <<\nlast\n>>\n";
    NSString *result = [group show];
    [self assertEquals:expected result:result];
}

- (void) testSingleTemplateWithArgs
{
    NSString *templates = @"t(a,b) ::= \"[<a>]\"\n";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    NSString *expected = @"t(a,b) ::= <<\n[<a>]\n>>\n";
    NSString *result = [group show];
    [self assertEquals:expected result:result];
}

- (void) testDefaultValues
{
    NSString *templates = @"t(a={def1},b=\"def2\") ::= \"[<a>]\"\n";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    NSString *expected = @"t(a={def1},b=\"def2\") ::= <<\n[<a>]\n>>\n";
    NSString *result = [group show];
    [self assertEquals:expected result:result];
}

- (void) testDefaultValueTemplateWithArg
{
    NSString *templates = @"t(a={x | 2*<x>}) ::= \"[<a>]\"\n";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    NSString *expected = @"t(a={x | 2*<x>}) ::= <<\n[<a>]\n>>\n";
    NSString *result = [group show];
    [self assertEquals:expected result:result];
}

- (void) testNestedTemplateInGroupFile
{
    NSString *templates = @"t(a) ::= \"<a:{x | <x:{y | <y>}>}>\"\n";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    NSString *expected = @"t(a) ::= <<\n<a:{x | <x:{y | <y>}>}>\n>>\n";
    NSString *result = [group show];
    [self assertEquals:expected result:result];
}

- (void) testNestedDefaultValueTemplate
{
    NSString *templates = @"t(a={x | <x:{y|<y>}>}) ::= \"ick\"\n";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    [group load];
    NSString *expected = @"t(a={x | <x:{y|<y>}>}) ::= <<\nick\n>>\n";
    NSString *result = [group show];
    [self assertEquals:expected result:result];
}

- (void) testNestedDefaultValueTemplateWithEscapes
{
    NSString *templates = @"t(a={x | \\< <x:{y|<y>\\}}>}) ::= \"[<a>]\"\n";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroup *group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    NSString *expected = @"t(a={x | \\< <x:{y|<y>\\}}>}) ::= <<\n[<a>]\n>>\n";
    NSString *result = [group show];
    [self assertEquals:expected result:result];
}

- (void) testMessedUpTemplateDoesntCauseRuntimeError
{
    NSString *templates = @"main(p) ::= <<\n<f(x=\"abc\")>\n>>\n\nf() ::= <<\n<x>\n>>\n";
    [self writeFile:tmpdir fileName:@"t.stg" content:templates];
    STGroupFile *group = nil;
    ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
    group = [[STGroupFile newSTGroupFile:[NSString stringWithFormat:@"%@/t.stg", tmpdir]] autorelease];
    [group setListener:errors];
    ST *st = [group getInstanceOf:@"main"];
    [st render];
    NSString *expected = @"[context [/main] 1:1 passed 1 arg(s) to template /f with 0 declared arg(s), context [/main] 1:1 attribute x isn't defined, context [/main /f] 1:1 attribute x isn't defined]";
    NSString *result = [errors.errors description];
    [self assertEquals:expected result:result];
    }
    
    @end
